File: app/src/main/java/net/nashihara/naroureader/adapters/SimpleRecyclerViewAdapter.java
package net.nashihara.naroureader.adapters;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import net.nashihara.naroureader.R;
import net.nashihara.naroureader.databinding.ItemSimpleRecyclerBinding;
import net.nashihara.naroureader.listeners.OnItemClickListener;
import net.nashihara.naroureader.models.entities.Novel4Realm;

import java.util.ArrayList;
import java.util.List;

public class SimpleRecyclerViewAdapter extends RecyclerView.Adapter<SimpleRecyclerViewAdapter.BindingHolder> {

    private LayoutInflater inflater;

    private ArrayList<Novel4Realm> arrayList;

    private OnItemClickListener listener;

    private RecyclerView recyclerView;

    public SimpleRecyclerViewAdapter(Context context) {
        this.inflater = LayoutInflater.from(context);
        arrayList = new ArrayList<>();
    }

    @Override
    public void onAttachedToRecyclerView(RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);
        this.recyclerView = recyclerView;
    }

    @Override
    public void onDetachedFromRecyclerView(RecyclerView recyclerView) {
        super.onDetachedFromRecyclerView(recyclerView);
        recyclerView = null;
    }

    @Override
    public BindingHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        final View v = inflater.inflate(R.layout.item_simple_recycler, parent, false);
        return new BindingHolder(v, listener);
    }

    @Override
    public void onBindViewHolder(BindingHolder holder, int position) {
        if (arrayList != null && arrayList.size() > position && arrayList.get(position) != null) {
            ItemSimpleRecyclerBinding binding = holder.getBinding();

            binding.title.setText(arrayList.get(position).getTitle());
        }
    }

    @Override
    public int getItemCount() {
        if (arrayList == null) {
            return 0;
        }
        return arrayList.size();
    }

    public void addDataOf(List<Novel4Realm> dataList) {
        int beforePos = arrayList.size();
        arrayList.addAll(dataList);
        notifyItemRangeInserted(beforePos, dataList.size());
    }

    public void clearData() {
        arrayList.clear();
    }

    public ArrayList<Novel4Realm> getList() {
        return this.arrayList;
    }

    public void setOnItemClickListener(OnItemClickListener listener) {
        this.listener = listener;
    }

    static class BindingHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        private ItemSimpleRecyclerBinding binding;
        private OnItemClickListener mListener;

        public BindingHolder(View itemView, OnItemClickListener listener) {
            super(itemView);
            this.mListener = listener;
            binding = DataBindingUtil.bind(itemView);

            binding.itemContainer.setOnClickListener(this);
        }

        public ItemSimpleRecyclerBinding getBinding(){
            return this.binding;
        }

        @Override
        public void onClick(View v) {
            if (mListener != null) {
                mListener.onItemClick(v, getLayoutPosition());
            }
        }
    }
}

File: app/src/main/java/net/nashihara/naroureader/adapters/SimpleRecyclerViewAdapter.kt
package net.nashihara.naroureader.adapters

import android.content.Context
import android.databinding.DataBindingUtil
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import net.nashihara.naroureader.R
import net.nashihara.naroureader.databinding.ItemSimpleRecyclerBinding
import net.nashihara.naroureader.entities.Novel4Realm

import java.util.ArrayList

class SimpleRecyclerViewAdapter(context: Context) : RecyclerView.Adapter<SimpleRecyclerViewAdapter.BindingHolder>() {

    private val inflater: LayoutInflater = LayoutInflater.from(context)

    val list = ArrayList<Novel4Realm>()

    private var listener: ((View, Int) -> Unit)? = null

    private var recyclerView: RecyclerView? = null

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView?) {
        super.onAttachedToRecyclerView(recyclerView)
        this.recyclerView = recyclerView
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView?) {
        var recyclerView = recyclerView
        super.onDetachedFromRecyclerView(recyclerView)
        recyclerView = null
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BindingHolder {
        val v = inflater.inflate(R.layout.item_simple_recycler, parent, false)
        return BindingHolder(v, listener)
    }

    override fun onBindViewHolder(holder: BindingHolder, position: Int) {
        if (list.size > position) {
            val binding = holder.binding

            binding.title.text = list[position].title
        }
    }

    override fun getItemCount(): Int {
        return list.size
    }

    fun addDataOf(dataList: List<Novel4Realm>) {
        val beforePos = list.size
        list.addAll(dataList)
        notifyItemRangeInserted(beforePos, dataList.size)
    }

    fun clearData() {
        list.clear()
    }

    fun setOnItemClickListener(listener: (View, Int) -> Unit){
      this.listener = listener
    }

    class BindingHolder(itemView: View, private val listener: ((View, Int) -> Unit)?) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        val binding: ItemSimpleRecyclerBinding = DataBindingUtil.bind<ItemSimpleRecyclerBinding>(itemView)

        init {
            binding.itemContainer.setOnClickListener(this)
        }

        override fun onClick(v: View) {
            listener?.invoke(v, layoutPosition)
        }
    }
}
